### ISSUE TITLE
Fail to load WebView View resources on Remote SSH even if asWebviewUri used

### ISSUE THREAD

**Issue by tamuratak** [microsoft/vscode#112641]
Fail to load WebView View resources on Remote SSH even if `asWebviewUri` used.

Version: 1.53.0-insider
Commit: f3e95ede9e4b9818c29a5bca3564d6fd162fb210
Date: 2020-12-15T08:14:20.855Z
Electron: 11.0.5
Chrome: 87.0.4280.88
Node.js: 12.18.3
V8: 8.7.220.29-electron.0
OS: Darwin x64 17.7.0

Steps to Reproduce:

1. Install [LaTeX Workshop](https://marketplace.visualstudio.com/items?itemName=James-Yu.latex-workshop) on the remote server with Remote SSH
2. Open a TeX file `t.tex`.
3. Fail to load WebView View resources
4. You do not have to install any TeX distribution.

![スクリーンショット 2020-12-16 17 51 21](https://user-images.githubusercontent.com/10665499/102326801-35968c80-3fc8-11eb-8470-6cc2d7fc34d0.png)

The console message of the WebView is the following:

```
Failed to load resource: the server responded with a status of 404 (Not Found) snippetview.css:1
Failed to load resource: the server responded with a status of 404 (Not Found) snippetview.js:1 
Failed to load resource: the server responded with a status of 404 (Not Found) snippetview.css:1 
```

The resource urls are the follwoing:

```
vscode-webview-resource://webviewview-latex-snippet-view/file///home/ubuntu/.vscode-server-insiders/extensions/james-yu.latex-workshop-8.15.1-beta.9/resources/snippetview/snippetview.css
vscode-webview-resource://webviewview-latex-snippet-view/file///home/ubuntu/.vscode-server-insiders/extensions/james-yu.latex-workshop-8.15.1-beta.9/resources/snippetview/snippetview.js
```

Through the dev tool, I confirm that these urls are actually used in the html of the WebView View.

I confirm that the resources exist:
```
$  ls -l /home/ubuntu/.vscode-server-insiders/extensions/james-yu.latex-workshop-8.15.1-beta.9/resources/snippetview/
total 36
-rw-r--r-- 1 ubuntu ubuntu  4751 Dec 16 08:44 snippetview.css
-rw-r--r-- 1 ubuntu ubuntu 21725 Dec 16 08:44 snippetview.html
-rw-r--r-- 1 ubuntu ubuntu  3374 Dec 16 08:44 snippetview.js
```
Does this issue occur when all extensions are disabled?: Yes

**Comment by tamuratak** [comment]
> If you see this again, can you try collapsing and then expanding the webview view. This should force the resources in the webview to be reloaded

Expanding and moving the views do not work. It still fails to load.

**Comment by tianjianchn** [comment]
I found a workaround for the WebviewView resources in Remote SSH:
```js
// example for css files
// cssFileContent = await fs.promises.readFile(cssFilePath, 'utf-8');
if(!!vscode.env.remoteName) { // run in remote mode
  return `<style>${cssFileContent}</style>`
}
else {
  return `<link href="${cssWebviewUrl}" rel="stylesheet">`
}
```
Same for js files.

**Comment by mjbvz** [comment]
@tamuratak Did you add a workaround for this issue? I don't see it in the latest insiders in an ssh workspace

**Comment by tamuratak** [comment]
No, we didn't. I still see the issue.

Version: 1.55.0-insider
Commit: c185983a683d14c396952dd432459097bc7f757f
Date: 2021-03-30T12:47:00.900Z
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Darwin x64 18.7.0

Name: LaTeX Workshop
Id: james-yu.latex-workshop
Version: 8.16.1

Name: Remote - SSH
Version: 0.65.1


![スクリーンショット 2021-03-31 8 55 18](https://user-images.githubusercontent.com/10665499/113071019-de2ace00-91fe-11eb-951e-2a5d056c056a.png)

**Comment by mjbvz** [comment]
Here's what I see, which I think is correct:

![Screen Shot 2021-03-30 at 5 16 22 PM](https://user-images.githubusercontent.com/12821956/113072330-a9296780-917b-11eb-8b9c-e61f43736711.png)

What type of remote are you connecting to?

**Comment by tamuratak** [comment]
> What type of remote are you connecting to?

Debian GNU/Linux bullseye/sid https://www.debian.org/releases/bullseye/

```
$  cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux bullseye/sid"
NAME="Debian GNU/Linux"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

$  uname -a
Linux 0d2d465213da 4.19.121-linuxkit #1 SMP Thu Jan 21 15:36:34 UTC 2021 x86_64 GNU/Linux
```

**Comment by LeuisKen** [comment]
I think I find the cause of this issue, it seems to be an race condition issue. Here is the call order in my case that can reproduce this problem. The code I'm using is tag 1.53.2:

1. First, `webviewService.createWebviewOverlay` method was called with `extension` argument is `undefined`;
![image](https://user-images.githubusercontent.com/7701575/114655820-8ff9ec80-9d1f-11eb-806d-aa21cd72f3e9.png)
Code here: https://github.com/microsoft/vscode/blob/1.53.2/src/vs/workbench/contrib/webviewView/browser/webviewViewPane.ts#L161

2. Then, the `_webviewManagerService.registerWebview` method was called with `extensionLocation` is `undefined`, the undefined extensionLocation is sent to the main process:
![image](https://user-images.githubusercontent.com/7701575/114657986-b3269b00-9d23-11eb-894c-ce07748f2672.png)
Code here: https://github.com/microsoft/vscode/blob/1.53.2/src/vs/workbench/contrib/webview/electron-sandbox/resourceLoading.ts#L83

3. At that time in main process, the undefined location registered in the webview metadata:
![image](https://user-images.githubusercontent.com/7701575/114658868-1fee6500-9d25-11eb-9356-e0fdf459bebe.png)
Code here: https://github.com/microsoft/vscode/blob/1.53.2/src/vs/platform/webview/electron-main/webviewProtocolProvider.ts#L113

4. Then the vscode-webview-resource request sent, and `loadLocalResource` function got undefined extensionLocation
![image](https://user-images.githubusercontent.com/7701575/114659116-8f645480-9d25-11eb-8d82-3553f42a9e77.png)
Code here: https://github.com/microsoft/vscode/blob/1.53.2/src/vs/platform/webview/electron-main/webviewProtocolProvider.ts#L212

5. The uri was not rewrite to http due to undefined extensionLocation:
![image](https://user-images.githubusercontent.com/7701575/114659295-ec600a80-9d25-11eb-8284-3871ffd858f2.png)
Code here: https://github.com/microsoft/vscode/blob/1.53.2/src/vs/platform/webview/electron-main/webviewProtocolProvider.ts#L167

7. After that
![image](https://user-images.githubusercontent.com/7701575/114658401-4eb80b80-9d24-11eb-98ec-aacf4109ac17.png)
Code here: https://github.com/microsoft/vscode/blob/1.53.2/src/vs/workbench/api/browser/mainThreadWebviewViews.ts#L82

If the step 7 happens before step 2, or the meta data update rightly, this bug will not reproduce.

And I find the pr #120654 change the webview resource loading strategy, I think maybe this bug is now fixed. : )

**Comment by tamuratak** [comment]
@LeuisKen Thanks.

I have confirmed that the issue has been fixed on VS Code Insiders 1.56.0.

Version: 1.56.0-insider
Commit: 394a2e4779ff767e2761a59a51b15848dcd24173
Date: 2021-04-13T11:11:34.416Z
Electron: 11.4.2
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Darwin x64 18.7.0

**Comment by mjbvz** [comment]
Thanks for the investigation @LeuisKen

Yes this should be fixed by #120654, which should be a more robust solution than the custom protocol we were  previously using to load resources in webviews
